### ISSUE TITLE
Support for QueryBus?

### ISSUE THREAD

**Issue by fmasa** [SimpleBus/MessageBus#52]
Current implementation of `MessageBusSupportingMiddleware` doesn't return result from handler.
Because of that it can't be used to create i.e. QueryBus.

Is there any reason for this design decision?

I know there was RFC already (#16), but it was closed.

There would be no BC break (as no return value means NULL right now).

**Comment by cmodijk** [comment]
@matthiasnoback the author of the SimpleBus Library's blogged about the reason to not return anything and the design of the package:
* http://php-and-symfony.matthiasnoback.nl/2015/01/a-wave-of-command-buses/
* http://php-and-symfony.matthiasnoback.nl/2015/01/responsibilities-of-the-command-bus/
* http://php-and-symfony.matthiasnoback.nl/2015/01/from-commands-to-events/
* http://php-and-symfony.matthiasnoback.nl/2015/01/some-questions-about-the-command-bus/
* http://php-and-symfony.matthiasnoback.nl/2015/01/collecting-events-and-the-events-aware-command-bus/

The short version is that its intentional because a Command Handler is not ment to return anything and  by not allowing this in the full stack you really can't and you should really think about command's and what the are doing. There was an discussed about implementing a query bus in #16 with some interesting suggestions but it was closed. Another issue on the subject of returning something is #47.

**Comment by fmasa** [comment]
AFAIK the package contains only universal message bus (which can be used as CommandBus, EventBus). I agree that command bus shouldn't return any result, but **CommandBus != QueryBus** as the purpose is different and "Command buses aren't supposed to do that" isn't the answer.

Many CQRS apps uses both command and query buses. Right now there is high duplicity necessary (= you have to implement query bus yourself or use different package).

**Comment by cmodijk** [comment]
Yes CommandBus != QueryBus but you question was 'Is there any reason for this design decision?' and there was thats what i tried to answer.  

In #16 @matthiasnoback decided to close the QueryBus part to get a better view on how te solve the question. I think it's create to start a discussion about this again because we where thinking on implementing QueryBus. Our main goal is to be able to send query commands and allow an async response by using for example RabbitMQ RPC. Our main purpose for this is that we need to query a closed system and by using RabbitMQ on the outside we could use a QueryBus with RPC to call internal job's te return the result.

**Comment by fmasa** [comment]
I see the problem. But in most cases queries are handled synchronously and for synchronous handling is return value best solution.

**Comment by cmodijk** [comment]
I agree we could stil use the return value if i implement the RPC method because it just waits until it's done. I don't want to query in async but use RabbitMQ as a communication method.

**Comment by matthiasnoback** [comment]
I'm not sure how often an actual query bus will make sense. When you need data/an object you ask a repository for it, and then you could implement the repository as a so-called facade, which behind-the-scenes actually does some RPC call.

I've come to think that the nature of a query bus is quite different as well. Both the command bus and the event bus can queue messages in-memory, then handle them one after another. See https://github.com/SimpleBus/MessageBus/blob/master/src/Bus/Middleware/FinishesHandlingMessageBeforeHandlingNext.php#L34. Supporting a return value in this example would be impossible since in fact there could be/are multiple return values.

Of course, you're free to implement a query bus. And it might be a nice package too. If any of you would like to work on it, I'd be happy to host it even at the SimpleBus organization. It should be really simple, but you can't use a lot from this library, except for the callable resolver stuff.

**Comment by cmodijk** [comment]
Hmm interesting the RPC inside the repository I had not thought of great idea tnx!

**Comment by fmasa** [comment]
@matthiasnoback I agree that `FinishesHandlingMessageBeforeHandlingNext` doesn't make sense when it comes to queries. I don't see problem with returning _NULL_ in that middleware. It will never be used in query handler, and support for return value for command handlers isn't desirable.

I was thinking about separate package, but it would contain only copy/paste from `MessageBusSupportingMiddleware` with two returns keyword.

**Comment by matthiasnoback** [comment]
@fmasa Right, totally get it. My point being, we can't just add those `return` statements in this package because it would be incompatible with the command bus/event bus (because of that `FinishesHandlingMessageBeforeHandlingNext` middleware - which return value will become *the* return value?). Besides, I feel that having no return value has so far actually been of great value in the "secret" purpose of pushing developers to learn CQRS :)

**Comment by MidnightDesign** [comment]
Just an idea: wouldn't it be possible to give the Query Bus an object that it should write its data to? That way it doesn't need to return anything.

**Comment by fmasa** [comment]
@MidnightDesign It's perfectly possible to "hack" it that way. But I still find it rather ugly. I'd like to have first-class support for query bus.

I partially agree with @matthiasnoback (point About forcing devs to learn about CQRS :) ), that it should be in separate package. The only thing I'm thinking about is how to avoid (small) duplicity of callable resolver.

**Comment by matthiasnoback** [comment]
Introducing a return value *is* in fact a BC break, since existing middlewares would have to return the value returned by `$next()`. This is why I don't want to introduce this change in the main package. If you want to do it, just create a new package (it can be an official SimpleBus package). And make it depend on the MessageBus package so you can reuse the callable/resolver stuff :)

**Comment by ajgarlag** [comment]
I've implemented a POC for a QueryBus based on the @MidnightDesign idea. You can review it at https://github.com/ajgarlag/AjglSimpleBusQueryBus

**Comment by warezgibzzz** [comment]
Hello, i've created a bundle for Symfony 2.x/3.x based on @ajgarlag implementation of QueryBus.
You can check it out here: [warezgibzzz/QueryBusBundle](https://github.com/warezgibzzz/QueryBusBundle)

**Comment by matthiasnoback** [comment]
I'm closing this issue for me to be able to keep an eye on what has to be done ;) Thanks for discussing this and offering your alternatives!
